src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -333,24 +336,33 @@ internal MappedField GetOrAddDataField(ImmutableArray<byte> data, ushort alignme
             }
 
             var @this = moduleBuilder.GetPrivateImplClass(syntaxNode, diagnostics);
-            return @this._dataSectionStringLiteralTypes.GetOrAdd(text, static (key, arg) =>
+            return @this._dataSectionStringLiteralTypes.GetOrAdd(text, static (text, arg) =>
             {
-                var (@this, data, diagnostics) = arg;
+                var (@this, data, syntaxNode, diagnostics) = arg;
 
-                string name = "<S>" + DataToHexViaXxHash128(data);
+                string name = "<S>" + @this.DataToHexViaXxHash128(data);
 
                 MappedField dataField = @this.GetOrAddDataField(data, alignment: 1);
 
                 Cci.IMethodDefinition bytesToStringHelper = @this.GetOrSynthesizeBytesToStringHelper(diagnostics);
 
+                var previousText = @this._dataSectionStringLiteralNames.GetOrAdd(name, text);
+                if (previousText != text)
+                {
+                    // If there is a hash collision, we cannot fallback to normal string literal emit strategy
+                    // because the selection of which literal would get which emit strategy would not be deterministic.
+                    var messageProvider = @this.ModuleBuilder.CommonCompilation.MessageProvider;
+                    diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_DataSectionStringLiteralHashCollision, syntaxNode.GetLocation(), previousText));
+                }
```

Comment:
>} [](http://example.com/codeflow?start=16&length=1)

`return null;`?

Reply:
Then the dictionary would contain `null` values, which would need to be handled elsewhere. I'd prefer to leave this as it is (note the emit would work even without the error, just the emitted binary would contain two types with the same name).